### DISCUSSION TITLE
new peer id connection failed

### DISCUSSION THREAD

**Discussion by RemonAhammad** [libp2p/rust-libp2p#4395] [Q&A]
Error : 
```
Dial error: condition Disconnected for dialing peer was false
```
connection code :
```
match swarm.dial(peer_id) {
        Ok(_) => handle_peer_connection().await,
        Err(e) => {
                    println!("Failed to connect new peer. id : {:?}\n", peer_id);
        }
    }
```
Behaviour:
```
pub struct MyBehaviour {
    pub gossipsub: gossipsub::Behaviour,
    pub mdns: mdns::async_io::Behaviour,
    pub reqres: libp2p::request_response::cbor::Behaviour<GreetRequest, GreetResponse>,
}

```
my swarm config:
```
pub fn configure_swarm(
    callback: EventStreamCallback,
) -> Result<
    (
        PeerId,
        Boxed<(PeerId, StreamMuxerBox)>,
        MyBehaviour,
        IdentTopic,
    ),
    Box<dyn Error>,
> {
    let id_keys = identity::Keypair::generate_ed25519();
    let local_peer_id = PeerId::from(id_keys.public());
    init_local_peerid_oncelock(local_peer_id.clone());
    send_peer_id_to_dart(callback, local_peer_id.clone().to_string());
    println!("Local PeerID: {}", local_peer_id);

    let noise_config = noise::Config::new(&id_keys)?;
    let trns = tcp::async_io::Transport::default()
        .upgrade(Version::V1)
        .authenticate(noise_config)
        .multiplex(yamux::Config::default())
        .boxed();

    let message_id_fn = |message: &gossipsub::Message| {
        let mut s = DefaultHasher::new();
        message.data.hash(&mut s);
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .expect("Failed to get time");
        let timestamp = now.as_secs() * 1000 + now.subsec_millis() as u64;
        let timestamp = timestamp.to_be_bytes();
        s.write(&timestamp);
        gossipsub::MessageId::from(s.finish().to_string())
    };

    let gossipsub_config = gossipsub::ConfigBuilder::default()
        .heartbeat_interval(Duration::from_secs(1))
        .validation_mode(gossipsub::ValidationMode::Strict)
        .message_id_fn(message_id_fn)
        .build()?;

    let mut gossipsub = gossipsub::Behaviour::new(
        gossipsub::MessageAuthenticity::Signed(id_keys),
        gossipsub_config,
    )?;
    let topic = gossipsub::IdentTopic::new("test-net");
    gossipsub.subscribe(&topic)?;

    let mdns = mdns::async_io::Behaviour::new(mdns::Config::default(), local_peer_id)?;

    let reqres = libp2p::request_response::cbor::Behaviour::<GreetRequest, GreetResponse>::new(
        [(
            libp2p::StreamProtocol::new("/my-cbor-protocol"),
            libp2p_request_response::ProtocolSupport::Full,
        )],
        libp2p::request_response::Config::default(),
    );

    let behaviour = MyBehaviour {
        gossipsub,
        mdns,
        reqres,
    };

    Ok((local_peer_id, trns, behaviour, topic))
}
```

**Comment by mxinden** [comment]
```
Dial error: condition Disconnected for dialing peer was false
```

By default libp2p only dials a peer if you are not already connected.

https://github.com/libp2p/rust-libp2p/blob/c6984dd51db6be8e260be1494f2bbc9052a42206/swarm/src/dial_opts.rs#L311-L325

The error message above implies, that you are already connected.

Does that apply in your case?

**Reply by RemonAhammad** [reply]
In my case, I just started the program, and that happened. This error happened when the first time this code was executed. It's occurred sometimes ( not every time).

**Comment by dariusc93** [comment]
The problem may lie in the fact that youre dialing a peer_id and not the address itself. Unless a behaviour is supplying the address of the peer, you will usually get that error because an address isnt supplied. 

EDIT: Also noticing the condition, if the peer is connected already, then it might error unless you change the condition too.

**Reply by mxinden** [reply]
> Unless a behaviour is supplying the address of the peer, you will usually get that error because an address isnt supplied.

In such case you should get a `DialError::NoAddresses`:

https://github.com/libp2p/rust-libp2p/blob/e36e8ce4456ff059dd54614913b521194d2d478c/swarm/src/lib.rs#L1543-L1544

**Comment by mxinden** [comment]
@blackshadowsoftwareltd I documented a possible improvement of the error in https://github.com/libp2p/rust-libp2p/issues/4407. Are you interested in contributing a fix?

**Reply by RemonAhammad** [reply]
sure. I was more debugging on this problem. I found the issue that when my app starts the 'SwarmEvent::ConnectionEstablished' event is triggered before 'SwarmEvent::Behaviour(MyBehaviourEvent::Mdns(mdns::Event::Discovered(list)))' this event. 

I made a new connection with a peer in 'Discovered' event. but there is a problem, the connection failed.

caused already the 'ConnectionEstablished' event triggered, the peer is already connected. that's why I got the panic.

**Comment by RemonAhammad** [comment]
I have another question. how much size of a message I can send to a peer using my Behaviour? 
I mean how much MB or Bytes?

**Reply by thomaseizinger** [reply]
For the cbor request-response abstraction, we have the following limits:

https://docs.rs/libp2p-request-response/0.25.1/src/libp2p_request_response/cbor.rs.html#55-58

There isn't a specific reason for these, we just needed an upper bound. Technically, we could make them configurable.
